### ISSUE TITLE
zOS support

### ISSUE THREAD

**Issue by VitekVlcek-Broadcom** [openmainframeproject/feilong#751]
It would be great if Feilong supported zOS lifecycle, where zOS is 2nd level zVM guest:

1. IPL zOS, 
2. Stop zOS 
3. invoke MVS commands (`D IPLINFO`, ...) and capture their response

My understanding is that Feilong runs a zlinux under zVM, what it is resource consumption of this image (average CPU utilization, Memory, Disk).  In my case I would ideally avoid extra VM with zlinux to reduce memory and CPU resources because I'm running zVM under zD&T (zPDT)

I already have a POC version implemented off the Feilong project here is an architecture diagram:
![zos and feilong](https://github.com/openmainframeproject/feilong/assets/43439912/8b2476a3-72ab-4253-b1b1-106d724670fa)
* In my case I can have many instances of zVM each representing a parallel sysplex, that could be provisioned by a developer, therefore I prefer to avoid zlinux on each zVM and instead have the majority of components off the zVM.
* Ideally, I would like to combine LXC and observer into a single VM to minimize overhead of infrastructure
* `LXC` is used to send commands such as:  `Q NAMES`, `CP SEND CP ZOS1 CP IPL ...`, `CP SEND CP ZOS1 VINPUT VMSG D IPLINFO`, `CP SET OBSERVER ...`
* `OBSERVER - IUCV2Socket` - capture console messages sent from zOS vms and forward them via a network socket to a CLI that runs on linux

**Comment by bjhuangr** [comment]
currently, feilong rely on the IUCV channel to communicate with SMAPI request servers, so it has to stay in the same z/VM system.
We don't have performance data like cpu consumption of feilong services, but according to our daily use, it's relatively a light weight service

**Comment by jichenjc** [comment]
from the picture I think you are proposing LXC and other CMS related components to hosting feilong like Linux
but provisioning z/OS Is not hard depend on the LXC related item, is that correct?

**Comment by jichenjc** [comment]
correct, that's what I am expecting about z/OS itself ,we can talk about at next meeting
if you can record/show the env that you have now it might be easier to understand

Thanks~
________________________________
From: Vitek Vlcek ***@***.***>
Sent: Thursday, October 12, 2023 8:14 PM
To: openmainframeproject/feilong ***@***.***>
Cc: CHEN JI ***@***.***>; Comment ***@***.***>
Subject: [EXTERNAL] Re: [openmainframeproject/feilong] zOS support (Issue #751)

LXC is used to issue CP commands Q NAMES, CP SEND CP ZOS1 CP IPL .. . , CP SEND CP ZOS1 VINPUT VMSG D IPLINFO, CP SET OBSERVER .. . from a linux VM running off the zVM. In this scenario, Feilong is not running in zLinux under zVM. This picture
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender
This message came from outside your organization.
<https://us-phishalarm-ewt.proofpoint.com/EWT/v1/PjiDSg!2A-qKrTUgROxke9zkqrvJs6yI8Q_iH1OjKqdnawhJKhS7JHeoCcjUMUMy5lLDLq1OvNLpVJdCC5YjG9howfYQK8Qi2FTSyHSfw0TABjigR_bAD65C6hd$>
Report Suspicious

ZjQcmQRYFpfptBannerEnd

LXC is used to issue CP commands Q NAMES, CP SEND CP ZOS1 CP IPL ..., CP SEND CP ZOS1 VINPUT VMSG D IPLINFO, CP SET OBSERVER ... from a linux VM running off the zVM. In this scenario, Feilong is not running in zLinux under zVM. This picture covers only the start/stop scenario of zOS. Provisioning of zOS is out of scope.
We could discuss late what it means to provision zOS it is not that easy like in case of linux, provisioning involves multiples volumes.

**Comment by mfriesenegger** [comment]
@VitekVlcek-Broadcom We mentioned this issue during today's Feilong TSC meeting.  I see @jichenjc suggested demoing the env.  Is this something you would like to do for the January Feilong meeting?

**Comment by Bischoff** [comment]
Hi Vitek, I have seen with interest your demo of your first PoC relying on LXC.

My observations:
 * I was first confused by "LXC" name, as it is a well known containers system under Linux :smile_cat: 
 * What you do with LXC could also be done via `s3270`, the "scripting" version of `x3270` terminal emulator: send CP commands, analyze output...
 * I don't think you need a Ubuntu host on your z/VM... You can probably directly use zthin from an outside host. If you do that, you lose the commodities of the REST server and are in direct grip with SMAPI.
 * I don't know if putting the Ubuntu host outside of the s/390 would work too.
 * Both suggestions above would fail if zthin needs IUCV to communicate with SMAPI. I think it's not the case and TCP/IP sockets are enough, but we are at the limits of my knowledge.
 * Your approach is very different in essence from what feilong does: you use CP command to start and stop an existing zVM guest, while feilong deploys a new VM from a preexisting image, by manipulating the z/VM directory via SMAPI.
 * The fact that the approach is different does not mean you cannot do the same thing - just in a different way :smile_cat: .
 
 I hope that helps.

**Comment by mfriesenegger** [comment]
Hello Eric,

> * I was first confused by "LXC" name, as it is a well known containers system under Linux 😸

I believe this LXC - https://www.vm.ibm.com/download/packages/descript.cgi?LXC - predates the Linux LXC technology that you are thinking of.

**Comment by jichenjc** [comment]
>>>I am quite confused whether zthin component of Feilong needs IUCV to communicate with SMAPI. @bjhuangr says above it does, but 1) I think SMAPI can also be driven with IP sockets, and 2) we accidentally got Feilong to work while we forgot IUCV settings 🤔 . If IUCV is not mandatory, then you could put zthin or even a full Feilong outside of the s390 host.

IUCV is used for Linux VM communication, SMAPI is for VM management through z/VM DIRMAINT/CPU
SMAPI is pure socket thing and feilong covers both SMAPI and IUCV for different purpose
you can create VM USER DIRECT through DIRAMINT but it might not fit feilong whole purpose which covers Linux part as well


>>>However, I do not know whether Feilong assumes it is a Linux system that is being deployed. For example you can access the IP address of the guest, I am not sure how this is achieved and if Feilong needs to access the guest OS to achieve that.

the access to IP address is beyong feilong scope, we used to use `ssh` to do management but this means this need management entity need ssh to VM which is some case it's security concern and sometimes the network is not reachable at all, so we use IUCV instead 


>>>If I were you, I would just try to start and stop an existing zOS system with Feilong, to see if that works.

if it's z/VM guest, I see no reason why those 2 are not supported (as it will be z/VM Start/Stop through Image_Activate and Deactivate API which is pure SMAPI thing which interpret as CP command only)

**Comment by Bischoff** [comment]
Thanks for the clarifications @jichenjc. The schema at https://cloudlib4zvm.readthedocs.io/en/latest/architecture.html#internal-architecture had mislead me, I suggest in another issue to fix it.

To summarize this discussion:
* handling z/OS VMs from Feilong would work as long as we use only the SMAPI part. That includes starting and stopping VMs.
* to do the same thing with z/OS as Feilong does with Linux, one would need to implement the equivalent functionality. The most "Feilongish" way would be IUCV, but you could continue with LXC, I guess
* but if you are to continue with LXC, is there still a point to rely on Feilong's SMAPI part? Yes, if provisioning becomes part of your scope.

These are just personal thoughts. I do not represent the Feilong project in any way.

**Comment by VitekVlcek-Broadcom** [comment]
Thanks @Bischoff for the summary, I would love to use SMAPI instead of LXC however. LXC is used for:

1. start/stop zVM - there should be API calls (based on your reply)
2. send a command to a VM console eg: `Q NAMES`, `CP SEND CP ZOS1 CP IPL ...`, `CP SEND CP ZOS1 VINPUT VMSG D IPLINFO`, `CP SET OBSERVER ...` - is there a SMAPI endpoint for it?

**Comment by jichenjc** [comment]
>>>start/stop zVM - there should be API calls (based on your reply)

yes, this is easier , Image_Activate/Deactivate I shared above

>>>send a command to a VM console eg: Q NAMES, CP SEND CP ZOS1 CP IPL ..., CP SEND CP ZOS1 VINPUT VMSG D IPLINFO, CP SET OBSERVER ... - is there a SMAPI endpoint for it?

we can create a customerized API doing so 
https://www.ibm.com/docs/en/zvm/7.3?topic=programming-creating-custom-apis

we used to have a API that can do so and it trigger some security concern as if something like 
`CP <input command>` then it's able to do everything which we don't want
